common/src/main/java/io/druid/math/expr/BuiltinFunctions.java

```diff
@@ -0,0 +1,809 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.math.expr;
+
+import io.druid.java.util.common.IAE;
+import org.joda.time.DateTime;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import java.util.List;
+
+/**
+ */
+public interface BuiltinFunctions extends Function.Library
+{
+
+  abstract class SingleParam implements Function
+  {
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      if (args.size() != 1) {
+        throw new IAE("function '%s' needs 1 argument", name());
+      }
+      Expr expr = args.get(0);
+      return eval(expr.eval(bindings));
+    }
+
+    protected abstract ExprEval eval(ExprEval param);
+  }
+
+  abstract class DoubleParam implements Function
+  {
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      if (args.size() != 2) {
+        throw new IAE("function '%s' needs 2 arguments", name());
+      }
+      Expr expr1 = args.get(0);
+      Expr expr2 = args.get(1);
+      return eval(expr1.eval(bindings), expr2.eval(bindings));
+    }
+
+    protected abstract ExprEval eval(ExprEval x, ExprEval y);
+  }
+
+  abstract class TripleParam implements Function
+  {
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      if (args.size() != 3) {
+        throw new IAE("function '%s' needs 3 arguments", name());
+      }
+      Expr expr1 = args.get(0);
+      Expr expr2 = args.get(1);
+      Expr expr3 = args.get(2);
+      return eval(expr1.eval(bindings), expr2.eval(bindings), expr3.eval(bindings));
+    }
+
+    protected abstract ExprEval eval(ExprEval x, ExprEval y, ExprEval z);
+  }
+
+  abstract class SingleParamMath extends SingleParam
+  {
+    @Override
+    protected final ExprEval eval(ExprEval param)
+    {
+      if (param.type() == ExprType.LONG) {
+        return eval(param.asLong());
+      } else if (param.type() == ExprType.DOUBLE) {
+        return eval(param.asDouble());
+      }
+      return ExprEval.of(null);
+    }
+
+    protected ExprEval eval(long param)
+    {
+      return eval((double) param);
+    }
+
+    protected ExprEval eval(double param)
+    {
+      return eval((long) param);
```

Comment:
Ah, I see this is just copy and pasted code. Nevermind. Still I think would be a nice change though.